extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -64,4 +179,216 @@ public boolean isOverwritable() {
             return false;
         }
     }
+
+    private abstract static class LazyImmutableContainer<T extends GenericContainer> implements GenericContainer {
+        private final byte[] serialized;
+        private final Schema schema;
+        private T deserialized;
+
+        protected LazyImmutableContainer(byte[] serialized, Schema schema) {
+            this.serialized = serialized;
+            this.schema = schema;
+        }
+
+        protected T deserialized() {
+            if (deserialized == null) {
+                deserialized = deserialize(new GenericDatumReader<>(schema), serialized);
+            }
+            return deserialized;
+        }
+
+        @Override
+        public Schema getSchema() {
+            return schema;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (!(obj instanceof GenericContainer
```

Comment:
`if (!(obj instanceof GenericContainer))` you can `return false` immediately, this will simplify condition and avoid useless `deserialized` invocation in some cases

Reply:
The second condition doesn't trigger deserialization, and it prevents unnecessary deserialization when the container types mismatch.

Reply:
true